cli/flox/src/utils/message.rs

```diff
@@ -47,3 +50,173 @@ pub(crate) fn package_installed(pkg: &PackageToInstall, environment_description:
         pkg.id()
     ));
 }
+
+/// A test spy for messages printed to stderr through the `message` module.
+/// In unit tests, the messaging functions of the `message` module will,
+/// populate a `History` in addition to printing the message to stderr.
+/// This allows the test to assert against the messages printed to stderr,
+/// without refactoring existing implementations to print to a `Write` trait
+/// and injecting a mock in tests.
+///
+/// The `History` is thus comparitivly non-intrusive.
+/// However, it comes with a few caveats:
+///
+/// Messaging functions are currently effectively (process) globals,
+/// inheriting the global nature of `stderr`.
+/// `#[test]` functions are run in parallel on separate _threads_
+/// by both `cargo test` and `cargo nextest`.
+/// Assuming an equally global `History`, messages from one test
+/// will be visible in another test.
+/// There are several approaches to avoid this:
+///
+/// * running tests in sequence (e.g. `cargo test -- --test-threads=1`),
+///   slows down the test suite.
+/// * additional bookkeeping of scopes (e.g. via `tracing`),
+///   complicates the test setup, wither needs to be done manually,
+///   or we maintain a `#[test]` replacement. (That said,
+///   there may be something to learn from `tracing`s architecture).
+/// * Pass a `History` instance to the messaging functions,
+///   intrusive as it reqeuries all callers to be updated.
+/// * Use a [thread_local!] `History`, which is the current approach.
+///
+/// Being thread local, the `History` is not shared between threads,
+/// i.e. not shared between tests.
+/// On the flip side, if the test spawns additional threads,
+/// which print messages, these will be captured in the `History`
+/// of the thread that spawned them, rather than the test's `History`.
+/// Likewise if the test uses a multi-threaded async runtime, like `tokio`,
+/// messages from spawned tasks will not be captured in the `History`.
+/// Also due to runtimes sharing threads between tasks, the `History` available
+/// in the task is also shared between tasks.
+///
+/// Luckily, our CLI surface is largely single threaded.
+/// Threads are mainly spawned for blocking operations either in the library layer,
+/// or to run library funcitons in the background, e.g. to print a progress bar
+/// on the main thread.
+/// Also, while technically within an async runtime, the CLI is largely synchronous,
+/// and the default [#[tokio::test]] runtime is single threaded.
+#[cfg(test)]
+pub mod history {
+    use std::cell::RefCell;
+    use std::collections::VecDeque;
+
+    thread_local! {
+        static THREAD_HISTORY: RefCell<History> = const {
+            RefCell::new(History {
+                messages:VecDeque::new(),
+            })
+        };
+    }
+
+    pub(crate) struct History {
```

Comment:
What's the point of having `History`? Seems like you can do essentially the same thing without it

<details>
<summary>Diff without History</summary>

```
diff --git a/cli/flox/src/utils/message.rs b/cli/flox/src/utils/message.rs
index b3de8d4d..cd426db9 100644
--- a/cli/flox/src/utils/message.rs
+++ b/cli/flox/src/utils/message.rs
@@ -8,7 +8,7 @@ use flox_rust_sdk::models::manifest::PackageToInstall;
 /// to include logging, word wrapping, ANSI filtereing etc.
 fn print_message(v: impl Display) {
     #[cfg(test)]
-    crate::utils::message::history::History::push_message(format!("{v}"));
+    crate::utils::message::history::push_message(format!("{v}"));

     eprintln!("{v}");
 }
@@ -100,56 +100,51 @@ pub mod history {
     use std::cell::RefCell;
     use std::collections::VecDeque;

+    type History = VecDeque<String>;
+
     thread_local! {
         static THREAD_HISTORY: RefCell<History> = const {
-            RefCell::new(History {
-                messages:VecDeque::new(),
-            })
+            RefCell::new(
+                VecDeque::new()
+            )
         };
     }

-    pub(crate) struct History {
-        messages: VecDeque<String>,
+    pub(crate) fn messages() -> VecDeque<String> {
+        THREAD_HISTORY.with_borrow(|h| h.clone())
     }

-    impl History {
-        pub(crate) fn messages() -> VecDeque<String> {
-            THREAD_HISTORY.with_borrow(|h| h.messages.clone())
-        }
-
-        pub(super) fn push_message(message: String) {
-            THREAD_HISTORY.with_borrow_mut(|global| {
-                global.messages.push_back(message);
-            });
-        }
+    pub(super) fn push_message(message: String) {
+        THREAD_HISTORY.with_borrow_mut(|global| {
+            global.push_back(message);
+        });
+    }

-        pub(crate) fn clear() {
-            THREAD_HISTORY.with_borrow_mut(|global| {
-                global.messages.clear();
-            });
-        }
+    pub(crate) fn clear() {
+        THREAD_HISTORY.with_borrow_mut(|global| {
+            global.clear();
+        });
     }

     /// Tests that show the behavior of the `History` in a concurrent environments.
     #[cfg(test)]
     mod tests {
-        use super::*;
-        use crate::utils::message::plain;
+        use crate::utils::message::{history, plain};

         #[test]
         fn concurrent_1() {
             plain("1");
-            assert_eq!(&History::messages(), &["1"])
+            assert_eq!(&history::messages(), &["1"])
         }

         #[test]
         fn concurrent_2() {
             plain("2");
-            assert_eq!(&History::messages(), &["2"])
+            assert_eq!(&history::messages(), &["2"])
         }

         async fn check_and_produce_inner(expect: &[&str]) {
-            assert_eq!(&History::messages(), &expect);
+            assert_eq!(&history::messages(), &expect);
             plain("async");
         }

@@ -157,14 +152,14 @@ pub mod history {
         async fn concurrent_async_3() {
             plain("3");
             check_and_produce_inner(&["3"]).await;
-            assert_eq!(&History::messages(), &["3", "async"])
+            assert_eq!(&history::messages(), &["3", "async"])
         }

         #[tokio::test]
         async fn concurrent_async_4() {
             plain("4");
             check_and_produce_inner(&["4"]).await;
-            assert_eq!(&History::messages(), &["4", "async"])
+            assert_eq!(&history::messages(), &["4", "async"])
         }

         /// This test will fail because the history is thread specific.
@@ -192,7 +187,7 @@ pub mod history {
                 plain("4");
             });
             let _ = tokio::join!(fut_1, fut_2, fut_3, fut_4);
-            assert_eq!(History::messages().len(), 4)
+            assert_eq!(history::messages().len(), 4)
         }

         /// This test will fail because the history is thread specific.
@@ -208,15 +203,15 @@ pub mod history {
                 scope.spawn(|| plain("4"));
             });

-            assert_eq!(History::messages().len(), 4)
+            assert_eq!(history::messages().len(), 4)
         }

         #[test]
         fn clear() {
             plain("message");
-            assert_eq!(&History::messages(), &["message"]);
-            History::clear();
-            assert_eq!(History::messages().len(), 0);
+            assert_eq!(&history::messages(), &["message"]);
+            history::clear();
+            assert_eq!(history::messages().len(), 0);
         }
     }
 }
```

</details>

Reply:
Ah, the History::* methods used to take (&mut self) in an earlier impl without refcell.
The `ThreadKey<refcell<History>>` type however doesnt give you a reference anymore at least not just like this so I made the methods static as well.
In the current impl i agree that could all be a static/free functions.

The benefit of wrapping them is that its easier to have two if we want to also track stderr without duplicating all methods for std*.
A struct also gives you something to implement helper methods on that can be tested,
things like "contains patterns".

If y'all prefer a grabbox of functions and potentially duplicating them your patch seems fine @mkenigs, otherwise i'd look if we can get a reference somehow

Reply:
I don't think this specific case is too important, but in general I don't want to add abstractions we don't need

That's a good point about stderr though, so I don't mind leaving it